clickstream-event-visualiser-ui/src/main/java/clickstream/event_visualiser_ui/data/datasource/CSEvDatasourceImpl.kt

```diff
@@ -0,0 +1,161 @@
+package clickstream.event_visualiser_ui.data.datasource
+
+import clickstream.event_visualiser.CSEVEventObserver
+import clickstream.event_visualiser_ui.data.model.CSEvEvent
+import clickstream.event_visualiser_ui.data.model.CSEvState
+import clickstream.interceptor.CSInterceptedEvent
+import kotlinx.coroutines.*
+import java.util.*
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.CopyOnWriteArrayList
+
+internal class CSEvDatasourceImpl constructor(private val evObserver: CSEVEventObserver) :
+    CSEvDatasource {
+
+    private val eventDataMap = ConcurrentHashMap<String, CopyOnWriteArrayList<CSEvEvent>>()
+    private var job: CompletableJob = SupervisorJob()
+    private var coroutineScope = CoroutineScope(job)
+
+    private val eventCallback: (List<CSInterceptedEvent>) -> Unit = {
+        coroutineScope.launch { addInterceptedEvent(it) }
+    }
+
+    override fun stopObserving() {
+        coroutineScope.cancel()
+        evObserver.removeCallback(eventCallback)
+    }
+
+    override fun startObserving() {
+        job = SupervisorJob()
+        coroutineScope = CoroutineScope(job)
+        evObserver.addCallback(eventCallback)
+    }
+
+    override suspend fun addEvents(eventList: List<CSEvEvent>) {
+
+    }
+
+    override suspend fun getAllEventNames(
+        keys: List<String>,
+        values: List<String>
+    ): List<String> = withContext(Dispatchers.Default) {
```

Comment:
the thread switching is better happening the invocation. which in the view model